src/jit/codegenarm64.cpp

```diff
@@ -450,6 +468,56 @@ void CodeGen::genBuildRegPairsStack(regMaskTP regsMask, ArrayStack<RegPair>* reg
         }
     }
     assert(regsCount == 0 && regsMask == RBM_NONE);
+
+    genSetUseSaveNextPairs(regStack);
+}
+
+//------------------------------------------------------------------------
+// genSetUseSaveNextPairs: Set useSaveNextPair for each RegPair on the stack which unwind info can be encoded as
+// save_next code.
+//
+// Arguments:
+//   regStack - a regStack instance to set useSaveNextPair.
+//
+// Notes:
+// We can use save_next for RegPair(N, N+1) only when we have sequence like (N-2, N-1), (N, N+1), (N+2, N+3).
+// In this case in the prolog save_next for (N, N+1) refers to save_pair(N-2, N-1),
+// in the epilog it refers to save_pair(N+2, N+3).
+//
+// static
+void CodeGen::genSetUseSaveNextPairs(ArrayStack<RegPair>* regStack)
+{
+    for (int i = 1; i < regStack->Height() - 1; ++i)
+    {
+        RegPair& curr = regStack->BottomRef(i);
+        RegPair  prev = regStack->Bottom(i - 1);
+        RegPair  next = regStack->Bottom(i + 1);
+
+        if (prev.reg2 == REG_NA || curr.reg2 == REG_NA || next.reg2 == REG_NA)
+        {
+            continue;
+        }
+
+        if (genCanUseSaveNextPair(prev, curr) && genCanUseSaveNextPair(curr, next))
```

Comment:
Is ` && genCanUseSaveNextPair(curr, next)` really required?

I would expect the code to not depend on `next`, so it would simply be:
```
    for (int i = 1; i < regStack->Height(); ++i)
    {
        RegPair& curr = regStack->BottomRef(i);
        RegPair& prev = regStack->BottomRef(i - 1);

        if (prev.reg2 == REG_NA || curr.reg2 == REG_NA)
        {
            continue;
        }

        if (genCanUseSaveNextPair(prev, curr))
        {
            curr.useSaveNextPair = true;
        }
    }
}
```

Reply:
Why can't we start epilog with `save_next`?

Reply:
fyi, you can see how `save_next` is handled (and how all the unwinding happens) here: https://github.com/dotnet/coreclr/blob/master/src/unwinder/arm64/unwinder_arm64.cpp

Reply:
Thank you for explaining this, fixed.